src/storage/src/render/upsert.rs

```diff
@@ -103,11 +105,82 @@ impl<H: Digest> Hasher for DigestHasher<H> {
     }
 }
 
+/// A tuple of the row or error, and the order by value.
+/// Will be used to store the state for a key in a hashmap in upsert
+#[derive(Clone, Debug, Hash, PartialEq, Eq, Serialize, Deserialize)]
+struct State(Result<Row, UpsertError>, UpsertOrder);
+
+/// Struct containing a list of values to order by
+///
+/// The following table shows the possible UpsertOrder values
+/// for a particular key with different value types where,
+/// - p = the UpsertOrder from previously persisted state
+/// - k = the UpsertOrder from incoming kafka ingestion
+///
+/// | value   | order by |p      |k      | k.cmp(p)                      |
+/// |---------|----------|-------|-------|-------------------------------|
+/// | Ok(row) | Y        |Some(a)|Some(b)| b.cmp(a)                      |
+/// | Err()   | Y        |None   |Some() | k > p always since Some > None|
+/// | Ok(row) | N        |None   |Some() | k > p always since Some > None|
+/// | Err()   | N        |None   |Some() | k > p always since Some > None|
+///
+#[derive(Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
+pub(crate) struct UpsertOrder(Option<SmallVec<[i128; 2]>>);
+
+impl From<MzOffset> for UpsertOrder {
+    fn from(val: MzOffset) -> Self {
+        let mut order = SmallVec::new();
+        order.push(
+            val.offset
+                .try_into()
+                .expect("Unexpected u64 overflow to i128"),
+        );
+        UpsertOrder(Some(order))
+    }
+}
+
+/// Gets the ordering information extracting out the value from the row
+/// based on the given index, and the default fallback value.
+/// Using smallvec with max size 2 as currently there can only be max two columns
+/// (timestamp and offset) to order things by
+pub(crate) fn get_order(
+    row_or_error: &Result<Row, UpsertError>,
+    order_by_indices: &SmallVec<[usize; 2]>,
```

Comment:
We don't need to know its a smallvec when we use it, so we can just take its `Deref` impl

```suggestion
    order_by_indices: &[usize],
```

Reply:
Oh cool!